File: java-to-kotlin/src/main/java/nl/alimeshkat/recipes/controller/RecipesController.java
package nl.alimeshkat.recipes.controller;

import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.alimeshkat.recipes.model.RecipeRequest;
import nl.alimeshkat.recipes.model.RecipeResponse;
import nl.alimeshkat.recipes.service.RecipesService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@AllArgsConstructor
@Slf4j
public class RecipesController {
    private final RecipesService recipeService;

    @PostMapping("recipes")
    public ResponseEntity<RecipeResponse> createRecipes(@Valid @RequestBody RecipeRequest recipeRequest) {
        log.info("Create Recipes {}", recipeRequest);
        return ResponseEntity.status(HttpStatus.CREATED).body(recipeService.saveRecipe(recipeRequest));

    }

    @GetMapping(value = "recipes")
    public ResponseEntity<Iterable<RecipeResponse>> getRecipes() {
        log.info("Get All Recipes");
        return ResponseEntity.ok(recipeService.getRecipes());
    }
    @GetMapping(value = "recipes/{id}")
    public ResponseEntity<RecipeResponse> getRecipe(@PathVariable() Long id) {
        log.info("Get Recipe with id: {}", id);
        return ResponseEntity.ok(recipeService.getRecipe(id));
    }
    @PutMapping(value = "recipes/{id}")
    public ResponseEntity<RecipeResponse> updateRecipe(@PathVariable() Long id, @Valid @RequestBody RecipeRequest recipeRequest) {
        log.info("Update Recipe with id: {}. Update: {}", id, recipeRequest);
        return ResponseEntity.ok(recipeService.updateOrCreateRecipe(id, recipeRequest));
    }
    @DeleteMapping("recipes/{id}")
    public ResponseEntity<Void> deleteRecipes(@PathVariable() Long id) {
        log.info("Delete Recipe {}", id);
        recipeService.deleteRecipe(id);
        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
    }


}

File: non-blocking-complete/src/main/kotlin/nl/alimeshkat/recipes/controller/RecipesController.kt
package nl.alimeshkat.recipes.controller

import jakarta.validation.Valid
import kotlinx.coroutines.flow.Flow
import nl.alimeshkat.recipes.model.RecipeRequest
import nl.alimeshkat.recipes.model.RecipeResponse
import nl.alimeshkat.recipes.service.RecipesService
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
class RecipesController(private val recipeService: RecipesService) {

    @PostMapping("recipes")
    suspend fun createRecipes(@Valid @RequestBody recipeRequest: RecipeRequest): ResponseEntity<RecipeResponse> {
        log.info("Create Recipes $recipeRequest")
        return ResponseEntity.status(HttpStatus.CREATED).body(
            recipeService.saveRecipe(recipeRequest)
        )
    }

    @GetMapping(value = ["recipes"])
    suspend fun recipes(): Flow<RecipeResponse> {
        log.info("Get All Recipes")
        return recipeService.getRecipes()
    }

    @GetMapping(value = ["recipes/{id}"])
    suspend fun getRecipe(@PathVariable id: Long): ResponseEntity<RecipeResponse> {
        log.info("Get Recipe with $id")
        return ResponseEntity.ok(recipeService.getRecipe(id))
    }

    @PutMapping(value = ["recipes/{id}"])
    suspend fun updateRecipe(
        @PathVariable id: Long,
        @Valid @RequestBody recipeRequest: RecipeRequest
    ): ResponseEntity<RecipeResponse> {
        log.info("Update Recipe with id: $id. Update: $recipeRequest")
        return ResponseEntity.ok(recipeService.updateOrCreateRecipe(id, recipeRequest))
    }

    @DeleteMapping("recipes/{id}")
    suspend fun deleteRecipes(@PathVariable id: Long): ResponseEntity<Void> {
        log.info("Delete Recipe $id")
        recipeService.deleteRecipe(id)
        return ResponseEntity.status(HttpStatus.NO_CONTENT).build()
    }

    companion object {
        val log: Logger = LoggerFactory.getLogger(RecipesController::class.java)
    }
}

